### PR TITLE
Move type check of MD array to managed

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Removes a HELPER_METHOD_FRAME.

Generally there are more of MD array can be moved to managed, but not completely now.

| Method        | Job        | Toolchain         | Mean     | Error     | StdDev    | Ratio | RatioSD |
|-------------- |----------- |------------------ |---------:|----------:|----------:|------:|--------:|
| SetCovariant  | Job-KZUIWY | \PR\corerun.exe   | 2.758 ns | 0.0252 ns | 0.0236 ns |  0.99 |    0.01 |
| SetCovariant  | Job-ZSCMRL | \main\corerun.exe | 2.780 ns | 0.0191 ns | 0.0169 ns |  1.00 |    0.00 |
|               |            |                   |          |           |           |       |         |
| SetExactMatch | Job-KZUIWY | \PR\corerun.exe   | 2.768 ns | 0.0323 ns | 0.0302 ns |  0.99 |    0.02 |
| SetExactMatch | Job-ZSCMRL | \main\corerun.exe | 2.795 ns | 0.0246 ns | 0.0230 ns |  1.00 |    0.00 |

Not over optimizing, just ensuring no regression.

### CODE REVIEW

File: src/coreclr/vm/corelib.h [hunk 1]
```diff
@@ -1169,6 +1165,9 @@ DEFINE_METHOD(CASTHELPERS, CHKCASTCLASSSPECIAL, ChkCastClassSpecial,      SM_Ptr
 DEFINE_METHOD(CASTHELPERS, UNBOX,            Unbox,                       SM_PtrVoid_Obj_RetRefByte)
 DEFINE_METHOD(CASTHELPERS, STELEMREF,        StelemRef,                   SM_ArrObject_IntPtr_Obj_RetVoid)
 DEFINE_METHOD(CASTHELPERS, LDELEMAREF,       LdelemaRef,                  SM_ArrObject_IntPtr_PtrVoid_RetRefObj)
+#ifdef FEATURE_ARRAYSTUB_AS_IL
```
Comment (huoyaoyuan) [inline src/coreclr/vm/corelib.h:1168]: It seems that we are ready to enable it for all platforms. Just separating to another PR.

**Comment by jkotas** [comment]
> Not over optimizing, just ensuring no regression.

Could you please share the source for the benchmark? It is a bit suspicious that the difference between the baseline and the PR is in the noise range. I would expect it to be a small measurable improvement.

File: src/coreclr/vm/corelib.h (hunk 1 above)
Comment (jkotas) [reply src/coreclr/vm/corelib.h:1168]: I agree it would be nice to enable this on all platforms and delete the #define.

**Review by jkotas — APPROVED** [review]
LGTM otherwise. Thank you!

**Comment by huoyaoyuan** [comment]
Updated test:

```csharp
        private Base[,] baseArray = new Base[10, 1];
        private Base[,] covariantArray = new Derived[10, 1];
        private Derived[,] derivedArray = new Derived[10, 1];

        private Derived derivedObj = new Derived();

        private void GenericSet<T>(T[,] array, T value)
        {
            for (int i = 0; i < 10; i++)
            {
                array[i, 0] = value;
            }
        }

        [Benchmark]
        public void ReferenceTypeBase() => GenericSet(baseArray, derivedObj);

        [Benchmark]
        public void ReferenceTypeExactMatch() => GenericSet(derivedArray, derivedObj);

        [Benchmark]
        public void ReferenceTypeCovariant() => GenericSet(covariantArray, derivedObj);
```

Result:
| Method                  | Job        | Toolchain         | Mean     | Error    | StdDev   | Ratio | RatioSD |
|------------------------ |----------- |------------------ |---------:|---------:|---------:|------:|--------:|
| ReferenceTypeBase       | Job-IECQRX | \PR\corerun.exe   | 32.91 ns | 0.183 ns | 0.171 ns |  0.76 |    0.01 |
| ReferenceTypeBase       | Job-OMBWNE | \main\corerun.exe | 43.08 ns | 0.219 ns | 0.194 ns |  1.00 |    0.00 |
|                         |            |                   |          |          |          |       |         |
| ReferenceTypeExactMatch | Job-IECQRX | \PR\corerun.exe   | 29.27 ns | 0.422 ns | 0.395 ns |  1.00 |    0.02 |
| ReferenceTypeExactMatch | Job-OMBWNE | \main\corerun.exe | 29.13 ns | 0.342 ns | 0.320 ns |  1.00 |    0.00 |
|                         |            |                   |          |          |          |       |         |
| ReferenceTypeCovariant  | Job-IECQRX | \PR\corerun.exe   | 28.59 ns | 0.229 ns | 0.214 ns |  1.00 |    0.01 |
| ReferenceTypeCovariant  | Job-OMBWNE | \main\corerun.exe | 28.68 ns | 0.181 ns | 0.169 ns |  1.00 |    0.00 |

Yes it is measuable improvement when `array.GetType().ElementType` isn't exact `obj.GetType()`. The static type of array doesn't matter here.

**Pull request merged**